### PR TITLE
codegen: ternary: Move allocations off stack

### PR DESCRIPTION
**Pull request by danobi**
See individual commits for more details.

Part of https://github.com/bpftrace/bpftrace/issues/3431.

##### Checklist

- [ ] Language changes are updated in `man/adoc/bpftrace.adoc`
- [ ] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

File: src/ast/passes/codegen_llvm.cpp [hunk 1]
```diff
@@ -1881,16 +1881,17 @@ void CodegenLLVM::visit(Ternary &ternary)
                                                "right",
                                                parent);
   BasicBlock *done = BasicBlock::Create(module_->getContext(), "done", parent);
+
+  Value *buf = nullptr;
+  if (ternary.type.IsStringTy()) {
```
Comment (jordalgo) [inline src/ast/passes/codegen_llvm.cpp:1886]: Maybe I'm missing something but why can't this logic be nested inside the `ternary.type.IsStringTy()` on line 1921?

Comment (danobi) [reply src/ast/passes/codegen_llvm.cpp:1886]: Yeah, I tried that initially but LLVM barfs during IR validation. It's b/c there's a CreateCondBr on line 1896, and if we put this initialization code before the insert points for left_block and right_block, we strand code between basic blocks. I wanted to keep the current code structure so I put the stuff up here

Comment (jordalgo) [reply src/ast/passes/codegen_llvm.cpp:1886]: Ah ok. Maybe just move this new code under the "// ordering of all the following statements is important" comment - as it seems to be important for this too

Comment (danobi) [reply src/ast/passes/codegen_llvm.cpp:1886]: Yeah good call

**Review by jordalgo — APPROVED** [review]

**Pull request merged**